albatross/core/model.h

```diff
@@ -13,266 +13,110 @@
 
-  virtual std::string get_name() const = 0;
+  template <typename T, typename FitModelType>
+  friend struct fit_type_from_fit_model_type;
 
-  virtual Insights get_insights() const { return insights_; }
+  template <typename T, typename FeatureType>
+  friend struct fit_type;
 
-  virtual void add_insights(const Insights &insights) {
-    for (const auto &insight : insights) {
-      insights_[insight.first] = insight.second;
-    }
-  };
-
-  virtual std::unique_ptr<RegressionModel<FeatureType>>
-  ransac_model(double inlier_threshold, std::size_t min_inliers,
-               std::size_t random_sample_size, std::size_t max_iterations) {
-    static_assert(
-        is_complete<GenericRansac<void, FeatureType>>::value,
-        "ransac methods aren't complete yet, be sure you've included ransac.h");
-    return make_generic_ransac_model<FeatureType>(
-        this, inlier_threshold, min_inliers, random_sample_size, max_iterations,
-        leave_one_out_indexer<FeatureType>);
-  }
+ private:
+  // Declaring these private makes it impossible to accidentally do things like:
+  //     class A : public ModelBase<B> {}
+  // or
+  //     using A = ModelBase<B>;
+  //
+  // which if unchecked can lead to some very strange behavior.
+  ModelBase() : insights_(){};
+  friend ModelType;
+  Insights insights_;
 
   /*
-   * Here we define the serialization routines.  Note that while in most
-   * cases we could use the cereal method `serialize`, in this case we don't
-   * know for sure where the parameters are stored.  The
-   * GaussianProcessRegression
-   * model, for example, derives its parameters from its covariance function,
-   * so it's `params_` are actually empty.  As a result we need to use the
-   * save/load cereal variant and deal with parameters through the get/set
-   * interface.
+   * Fit
    */
-  template <class Archive> void save(Archive &archive) const {
-    auto params = get_params();
-    archive(cereal::make_nvp("parameters", params));
-    archive(cereal::make_nvp("has_been_fit", has_been_fit_));
-  }
-
-  template <class Archive> void load(Archive &archive) {
-    auto params = get_params();
-    archive(cereal::make_nvp("parameters", params));
-    archive(cereal::make_nvp("has_been_fit", has_been_fit_));
-    set_params(params);
-  }
+  template <
+      typename FeatureType,
+      typename std::enable_if<has_valid_fit<ModelType, FeatureType>::value,
+                              int>::type = 0>
+  auto
+  fit_(const std::vector<FeatureType> &features,
+       const MarginalDistribution &targets) const {
+    const auto fit = derived().fit(features, targets);
+    return FitModel<ModelType, decltype(fit)>(derived(), std::move(fit));
+  }
+
+  template <typename FeatureType,
+            typename std::enable_if<
+                    has_possible_fit<ModelType, FeatureType>::value &&
+                    !has_valid_fit<ModelType, FeatureType>::value,
+                int>::type = 0>
+  FitModel<ModelType, FeatureType>
```

Comment:
At this point, the only thing CRTP is really adding is the ability to inspect whether or not the user has properly defined `fit` and `predict` in their implementations.  Eventually I plan on adding more informative inspection to make sure `fit` and `predict` methods work together, etc ...

Reply:
If I'm understanding you right, you'd like to allow users to write a single `predict` method and for Albatross to extract the return type and use that as a proxy for `PredictTypeIdentity<T>` when selecting which function to use? That should be possible, and pretty snazzy!

Reply:
Yeah exactly!  In some situations you want to be able to specialize method for each `PredictType`.  For example, you may have a much more efficient way of returning only the mean value instead of a full `JointDistribution` in which case using a `PredictIdentity` in the signature is (I think) unavoidable.  But for a basic model implementation you may not want to worry about that, in which case we could infer the `PredictIdentity`.